### PR TITLE
BACS: Input form scene

### PR DESCRIPTION
**Pull request by nauaros**
## Summary

This PR contains the work to build the BACS Input form scene.

Dev steps:
- [x] Form UI
- [x] Validation
- [x] Content
- [x] Unit tests

### CODE REVIEW

File: AdyenComponents/BACS Direct Debit/Factories/BACSDirectDebitItemsFactory.swift [hunk 1]
```diff
@@ -0,0 +1,170 @@
+//
+// Copyright (c) 2021 Adyen N.V.
+//
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
+//
+
+import Adyen
+import Foundation
+
+internal protocol BACSDirectDebitItemsFactoryProtocol {
```
Comment (erenbesel) [inline AdyenComponents/BACS Direct Debit/Factories/BACSDirectDebitItemsFactory.swift:10]: What is the intention of this protocol exactly? It does not seem like a reusable interface or a type constraint so curious what it accomplishes

File: AdyenComponents/BACS Direct Debit/Scenes/Input/BACSDirectDebitInputFormView.swift [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+import Adyen
+import Foundation
+
+internal protocol BACSDirectDebitInputFormViewProtocol {
+    func add<T: FormItem>(item: T?)
+    func displayValidation()
+}
+
+internal class BACSDirectDebitInputFormView: FormViewController, BACSDirectDebitInputFormViewProtocol {
```
Comment (erenbesel) [inline AdyenComponents/BACS Direct Debit/Scenes/Input/BACSDirectDebitInputFormView.swift:15]: Naming a controller as view can be confusing later on. We treat them as the same time to time but yeah :)

File: AdyenComponents/BACS Direct Debit/Scenes/Input/BACSDirectDebitInputFormView.swift [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    // MARK: - BACSDirectDebitInputFormViewProtocol
+
+    internal func add<T: FormItem>(item: T?) {
+        guard let item = item else { return }
+        append(item)
+    }
+
+    internal func displayValidation() {
```
Comment (erenbesel) [inline AdyenComponents/BACS Direct Debit/Scenes/Input/BACSDirectDebitInputFormView.swift:39]: There is `showValidation()` in FormViewController that's called in `validate()` already.

That won't work for this one? 2 reasons to ask this:  
- In the continuepayment method, calling `validate()` should ideally be enough without needing `isFormValid` call separately.
- `_ = validate()` is mildly annoying

File: AdyenComponents/BACS Direct Debit/Scenes/Input/BACSDirectDebitInputPresenter.swift [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        view.add(item: FormSpacerItem(numberOfSpaces: 2))
+        view.add(item: amountConsentToggleItem)
+        view.add(item: FormSpacerItem(numberOfSpaces: 1))
+        view.add(item: legalConsentToggleItem)
+        view.add(item: FormSpacerItem(numberOfSpaces: 2))
+        view.add(item: continueButtonItem)
+        view.add(item: FormSpacerItem(numberOfSpaces: 1))
+    }
+
+    private var isFormValid: Bool {
```
Comment (erenbesel) [inline AdyenComponents/BACS Direct Debit/Scenes/Input/BACSDirectDebitInputPresenter.swift:77]: Maybe call `validate()` in this function instead of `displayValidation` and remove that one?

File: AdyenComponents/BACS Direct Debit/Scenes/Input/BACSDirectDebitInputFormView.swift (hunk 2 above)
Comment (nauaros) [reply AdyenComponents/BACS Direct Debit/Scenes/Input/BACSDirectDebitInputFormView.swift]: Renamed to `BACSDirectDebitInputFormViewController`

File: Adyen/Assets/Generated/LocalizationKey.swift [hunk 5]
```diff
@@ -2,7 +2,7 @@
 // Copyright (c) 2021 Adyen N.V.
 //
 // This file is open source and available under the MIT license. See the LICENSE file for more info.
-// This file is autogenerated. Please do not modify it.
```
Comment (descorp) [inline Adyen/Assets/Generated/LocalizationKey.swift:5]: ❓

Comment (nauaros) [reply Adyen/Assets/Generated/LocalizationKey.swift:5]: Ups, thanks. I will restore it

File: Adyen/UI/Form/Items/Toggle/FormToggleItemView.swift [hunk 6]
```diff
@@ -9,7 +9,33 @@ import UIKit
 /// A view representing a switch item.
 /// :nodoc:
 public final class FormToggleItemView: FormValueItemView<Bool, FormToggleItemStyle, FormToggleItem> {
-    
+
+    // MARK: - UI elements
+
+    private lazy var stackView: UIStackView = {
```
Comment (descorp) [inline Adyen/UI/Form/Items/Toggle/FormToggleItemView.swift:15]: I thought this was done in another PR?

Comment (nauaros) [reply Adyen/UI/Form/Items/Toggle/FormToggleItemView.swift:15]: Yes, the fix is on dev already.

File: Adyen/Validators/BACSAccountNumberValidator.swift [hunk 7]
```diff
@@ -0,0 +1,9 @@
+//
+//  BACSAccountNumberValidator.swift
+//  Adyen
+//
+//  Created by Naufal Aros on 11/19/21.
+//  Copyright © 2021 Adyen. All rights reserved.
+//
+
+import Foundation
```
Comment (descorp) [inline Adyen/Validators/BACSAccountNumberValidator.swift:9]: Empty file

Comment (nauaros) [reply Adyen/Validators/BACSAccountNumberValidator.swift]: removed, thanks

File: AdyenDropIn/DropInComponent.swift [hunk 8]
```diff
@@ -218,6 +218,8 @@ public final class DropInComponent: NSObject, PresentableComponent {
         component.payment = configuration.payment
         
         switch component {
+        case let component as BACSDirectDebitComponent:
```
Comment (descorp) [inline AdyenDropIn/DropInComponent.swift:221]: Maybe we can do something about this 🤔
Some protocol maybe?

File: Adyen/UI/Form/FormViewController+ViewProtocol.swift [hunk 9]
```diff
@@ -0,0 +1,24 @@
+//
+// Copyright (c) 2021 Adyen N.V.
+//
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
+//
+
+import Foundation
+
+public protocol FormViewProtocol {
```
Comment (mohammedDehairy) [inline Adyen/UI/Form/FormViewController+ViewProtocol.swift:9]: no documentation? can you add `/// :nodoc:`

File: AdyenComponents/BACS Direct Debit/BACSDirectDebitComponent.swift [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+/// A component that provides a form for BACS Direct Debit payments.
+public final class BACSDirectDebitComponent: PaymentComponent, PresentableComponent {
+
+    // MARK: - PresentableComponent
+
+    public var viewController: UIViewController
+
+    public var requiresModalPresentation: Bool = false
```
Comment (mohammedDehairy) [inline AdyenComponents/BACS Direct Debit/BACSDirectDebitComponent.swift:23]: can you add `/// :nodoc:` to both properties?

File: AdyenComponents/BACS Direct Debit/BACSDirectDebitComponent.swift [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        self.style = style
+        self.localizationParameters = localizationParameters
+
+        let view = BACSDirectDebitInputFormViewController(title: paymentMethod.name,
+                                                          styleProvider: style)
+        self.viewController = view as UIViewController
+
+        let itemsFactory = BACSDirectDebitItemsFactory(styleProvider: style,
+                                                       localizationParameters: localizationParameters,
+                                                       scope: self)
```
Comment (mohammedDehairy) [inline AdyenComponents/BACS Direct Debit/BACSDirectDebitComponent.swift:63]: I see `scope` is retained by `BACSDirectDebitItemsFactory`, and `itemsFactory` is retained by the `presenter` and presenter is retained by self which is the `scope`, this is a retain cycle, please make the `scope` weak in `BACSDirectDebitItemsFactory`.

File: AdyenComponents/BACS Direct Debit/BACSDirectDebitComponent.swift [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+// MARK: - BACSDirectDebitRouterProtocol
+
+extension BACSDirectDebitComponent: BACSDirectDebitRouterProtocol {
+
+    func presentConfirmation(with data: BACSDirectDebitData) {
+        // TODO: - Continue payment logic
+        // 1. Assamble confirmation scene
+        // 2. Present confirmation scene
+        print("PAYMENT: \(data)")
```
Comment (mohammedDehairy) [inline AdyenComponents/BACS Direct Debit/BACSDirectDebitComponent.swift:79]: please use `adyenPrint` instead of swift `print`

File: AdyenComponents/BACS Direct Debit/BACSDirectDebitComponent.swift [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+    func confirmPayment(with data: BACSDirectDebitData) {
+        // TODO: - Payment processing logic
+        guard let bacsDirectDebitPaymentMethod = paymentMethod as? BACSDirectDebitPaymentMethod else {
+            return
+        }
+        let bacsDirectDebitDetails = BACSDirectDebitDetails(paymentMethod: bacsDirectDebitPaymentMethod,
+                                                            holderName: data.holderName,
+                                                            bankAccountNumber: data.bankAccountNumber,
+                                                            bankLocationId: data.bankLocationId)
```
Comment (mohammedDehairy) [inline AdyenComponents/BACS Direct Debit/BACSDirectDebitComponent.swift:95]: you create the `bacsDirectDebitDetails ` and do nothing with it?

File: AdyenComponents/BACS Direct Debit/BACSDirectDebitComponent.swift [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+        let presenter = BACSDirectDebitPresenter(view: view,
+                                                 router: self,
+                                                 itemsFactory: itemsFactory)
+        view.presenter = presenter
+    }
+}
+
+// MARK: - BACSDirectDebitRouterProtocol
+
+extension BACSDirectDebitComponent: BACSDirectDebitRouterProtocol {
```
Comment (mohammedDehairy) [inline AdyenComponents/BACS Direct Debit/BACSDirectDebitComponent.swift:73]: is this extension `internal`? can you add `internal` to its definition?

File: AdyenComponents/BACS Direct Debit/Scenes/Input/BACSDirectDebitInputFormViewController.swift [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+    @objc
+    internal func didTapCancelButton() {
+        resignFirstResponder()
+        presenter?.didCancel()
+    }
+}
+
+extension BACSDirectDebitInputFormViewController: ViewControllerDelegate {
+    func viewDidLoad(viewController: UIViewController) {
+        presenter?.viewDidLoad()
```
Comment (mohammedDehairy) [inline AdyenComponents/BACS Direct Debit/Scenes/Input/BACSDirectDebitInputFormViewController.swift:46]: you need to call the analytics event here, right?

File: Adyen/UI/Form/Items/Toggle/FormToggleItemView.swift (hunk 6 above)
Comment (descorp) [reply Adyen/UI/Form/Items/Toggle/FormToggleItemView.swift:15]: I would advise updating to latest `develop` then 😆

File: AdyenComponents/BACS Direct Debit/BACSDirectDebitComponent.swift (hunk 11 above)
Comment (descorp) [reply AdyenComponents/BACS Direct Debit/BACSDirectDebitComponent.swift:63]: `scope` could be just String.

File: Tests/AdyenTests/Adyen Tests/Components/BACS Direct Debit/Scenes/Input/BACSDirectDebitPresenterTests.swift [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+        // Then
+        XCTAssertEqual(router.presentConfirmationWithDataCallsCount, 0)
+    }
+
+    func testContinuePaymentWhenItemsAreValidShouldCallRouterPresentConfirmation() throws {
+        // Given
+        sut.amountConsentToggleItem?.value = true
+        sut.legalConsentToggleItem?.value = true
+
+        sut.holderNameItem?.value = "Katrina del Mar"
```
Comment (descorp) [inline Tests/AdyenTests/Adyen Tests/Components/BACS Direct Debit/Scenes/Input/BACSDirectDebitPresenterTests.swift:147]: we can add all this into `Dummy.shopper`, so you don't have to copy it 🤔

File: AdyenComponents/BACS Direct Debit/BACSDirectDebitComponent.swift (hunk 10 above)
Comment (nauaros) [reply AdyenComponents/BACS Direct Debit/BACSDirectDebitComponent.swift:25]: Added, thanks!

File: AdyenComponents/BACS Direct Debit/BACSDirectDebitComponent.swift (hunk 12 above)
Comment (nauaros) [reply AdyenComponents/BACS Direct Debit/BACSDirectDebitComponent.swift]: Changed to use `adyenPrint`

File: AdyenComponents/BACS Direct Debit/BACSDirectDebitComponent.swift (hunk 13 above)
Comment (nauaros) [reply AdyenComponents/BACS Direct Debit/BACSDirectDebitComponent.swift:97]: The next PR will use it to submit the payment transaction.

File: AdyenComponents/BACS Direct Debit/BACSDirectDebitComponent.swift (hunk 14 above)
Comment (nauaros) [reply AdyenComponents/BACS Direct Debit/BACSDirectDebitComponent.swift:75]: The `BACSDirectDebitRouterProtocol` is declared with access level `internal` already. So any conformance to it will be at that access level. Also, I believe it's not possible to declare `internal` extensions that declare protocol conformances.

File: AdyenComponents/BACS Direct Debit/Scenes/Input/BACSDirectDebitInputFormViewController.swift (hunk 15 above)
Comment (nauaros) [reply AdyenComponents/BACS Direct Debit/Scenes/Input/BACSDirectDebitInputFormViewController.swift:46]: Yes, indeed. Let me add this as a task for the following PR. Thanks for noticing :)

File: Tests/AdyenTests/Adyen Tests/Components/BACS Direct Debit/Scenes/Input/BACSDirectDebitPresenterTests.swift (hunk 16 above)
Comment (nauaros) [reply Tests/AdyenTests/Adyen Tests/Components/BACS Direct Debit/Scenes/Input/BACSDirectDebitPresenterTests.swift:147]: That's a good idea Vlad! I will add Katrina as a dummy shopper then.

File: AdyenComponents/BACS Direct Debit/BACSDirectDebitComponent.swift (hunk 11 above)
Comment (nauaros) [reply AdyenComponents/BACS Direct Debit/BACSDirectDebitComponent.swift:65]: `scope` (BACSComponent) doesn't retain the items factory, only used on init but it doesn't keep a reference to it.

I was checking the dependencies. Since the `BACSView` is weakly retaining `BACSPresenter`, there's no retain cycle. Here's a diagram for better clarity.

Screen Shot 2021-12-06 at 11 38 56

Despite all this, I agree with @descorp. I'm gonna use scope as String instead. What do you think?

Comment (nauaros) [reply AdyenComponents/BACS Direct Debit/BACSDirectDebitComponent.swift]: Used String instead

File: Adyen/UI/Form/FormViewController+ViewProtocol.swift (hunk 9 above)
Comment (nauaros) [reply Adyen/UI/Form/FormViewController+ViewProtocol.swift:10]: done

File: AdyenComponents/BACS Direct Debit/BACSDirectDebitComponent.swift (hunk 14 above)
Comment (nauaros) [reply AdyenComponents/BACS Direct Debit/BACSDirectDebitComponent.swift:75]: Marked functions as `internal` instead.

File: Tests/AdyenTests/Adyen Tests/Components/BACS Direct Debit/Scenes/Input/BACSDirectDebitPresenterTests.swift (hunk 16 above)
Comment (nauaros) [reply Tests/AdyenTests/Adyen Tests/Components/BACS Direct Debit/Scenes/Input/BACSDirectDebitPresenterTests.swift]: Refactored to use a mock variable.

**Review by mohammedDehairy — APPROVED** [review]

File: AdyenDropIn/DropInComponent.swift [hunk 17]
```diff
@@ -238,7 +240,7 @@ public final class DropInComponent: NSObject, PresentableComponent {
         
         if isRoot {
             self.delegate?.didFail(with: ComponentError.cancelled, from: self)
-        } else {
+        } else if component.requiresModalPresentation {
```
Comment (erenbesel) [inline AdyenDropIn/DropInComponent.swift:243]: Don't we need to add an 'else' case still?

Comment (nauaros) [reply AdyenDropIn/DropInComponent.swift:243]: I don't think so cause we only pop on the navigation controller if the component is presented modally

Comment (descorp) [reply AdyenDropIn/DropInComponent.swift:243]: So, when cancelling non-modal presentation (like SFViewController, PKApplePayViewConrtoller etc) `requiresModalPresentation` is clicked we are going to do nothing?

Comment (nauaros) [reply AdyenDropIn/DropInComponent.swift]: After off chat call with @descorp, reverting this change.

**Review by descorp — APPROVED** [review]
